File: 剑指offer/10-fibonacci数列/jumpFloor.cpp
class Solution {
public:
    int jumpFloor(int number) {
        if (number <= 0)
            return 0;
        else if (number == 1 || number == 2)
            return number;
        int fn_1 = 2, fn_2 = 1; // f(n-1) 与 f(n-2)
        int fn;
        for (int i = 3; i <= number; ++i)
        {
            fn = fn_1 + fn_2;  // f(n) = f(n-1) + f(n-2)
            fn_2 = fn_1;
            fn_1 = fn;
        }
        return fn;
    }
};

File: 剑指offer/10-fibonacci数列/jumpFloor.py
# -*- coding:utf-8 -*-
"""
一只青蛙一次可以跳上1级台阶，也可以跳上2级。
求该青蛙跳上一个n级的台阶总共有多少种跳法（先后次序不同算不同的结果）。
"""


class Solution:
    def jumpFloor(self, n):
        """
        提示：
        设 n 级台阶有 f(n) 种跳法，青蛙最后一次跳，有两种跳法：
            若最后一次跳只跳一个台阶则前面的 n-1 级有 f(n-1) 种方法
            若最后一次跳只跳两个台阶则前面的 n-2 级有 f(n-2) 种方法
        """
        # write code here

        f0 = 0
        f1 = 1
        f2 = 0
        if n == 0:
            return f0
        elif n == 1:
            return f1
        else:
            for i in range(2, n + 1):
                f2 = f1 + f0
                f1, f0 = f2, f1
            return f2

